### PR TITLE
[MU4] Network manager

### PR DESCRIPTION
**Pull request by Eism**

### CODE REVIEW

File: framework/network/inetworkmanager.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    virtual ~INetworkManager() = default;
+
+    virtual Ret get(const QUrl& url, QIODevice* incommingData) = 0;
+
+    virtual async::Channel<Progress> downloadProgressChannel() const = 0;
+
+    virtual void abort() = 0;
+};
+
+typedef QSharedPointer<INetworkManager> INetworkManagerPtr;
```
Comment (igorkorsukov) [inline framework/network/inetworkmanager.h:44]: Please use `std::shared_ptr`

File: framework/network/internal/networkmanager.cpp [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    IF_ASSERT_FAILED(reply) {
+        return;
+    }
+
+    m_incommingData->write(reply->readAll());
+}
+
+void NetworkManager::onDownProgress(qint64 current, qint64 total)
+{
+    emit downloadProgress(current, total);
```
Comment (igorkorsukov) [inline framework/network/internal/networkmanager.cpp:104]: Why not just send data to the `m_downloadProgressCh` channel?

File: framework/network/networkerrors.h [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+namespace mu {
+namespace framework {
+enum class Err {
+    Undefined       = int(Ret::Code::Undefined),
+    NoError         = int(Ret::Code::Ok),
+    UnknownError    = int(Ret::Code::NetworkFirst),
+
+    Abort,
+    Timeout,
+    Network,
```
Comment (igorkorsukov) [inline framework/network/networkerrors.h:34]: Better `NetworkError`

File: framework/network/networkerrors.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    HostClosed,
+    HostNotFound,
+    OpenIODeviceRead,
```
Comment (igorkorsukov) [inline framework/network/networkerrors.h:37]: Better `FailedOpenIODeviceToRead`

File: framework/network/networkerrors.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    OpenIODeviceWrite
```
Comment (igorkorsukov) [inline framework/network/networkerrors.h:38]: Better `FailedOpenIODeviceToWrite`

File: framework/network/inetworkmanager.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#ifndef MU_FRAMEWORK_INETWORKMANAGER_H
+#define MU_FRAMEWORK_INETWORKMANAGER_H
+
+#include <QUrl>
+#include <QIODevice>
```
Comment (RomanPudashkin) [inline framework/network/inetworkmanager.h:23]: class QUrl;
class QIODevice;

File: framework/network/inetworkmanagercreator.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+namespace mu {
+namespace framework {
+class INetworkManagerCreator : MODULE_EXPORT_INTERFACE
+{
+    INTERFACE_ID(INetworkManagerCreator)
+
+public:
+    virtual ~INetworkManagerCreator() = default;
+
+    virtual INetworkManagerPtr newNetworkManager() = 0;
```
Comment (RomanPudashkin) [inline framework/network/inetworkmanagercreator.h:34]: make?

File: framework/network/networktypes.h [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#ifndef MU_FRAMEWORK_NETWORKTYPES_H
+#define MU_FRAMEWORK_NETWORKTYPES_H
+
+namespace mu {
+namespace framework {
+struct Progress {
+    qint64 current = 0;
```
Comment (RomanPudashkin) [inline framework/network/networktypes.h:25]: int64_t

File: framework/network/internal/networkmanager.h [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    virtual ~NetworkManager() override;
+
+    Ret get(const QUrl& url, QIODevice* incommingData) override;
+
+    async::Channel<Progress> downloadProgressChannel() const override;
+
+    void abort() override;
+
+signals:
+    void downloadProgress(qint64 current, qint64 total);
```
Comment (igorkorsukov) [inline framework/network/internal/networkmanager.h:43]: This signal is no longer needed

**Pull request merged**